Simplified/AxisService/NYPLAxisTaskAggregator.swift

```diff
@@ -0,0 +1,178 @@
+  /// license.json does not contain the required keys for validating
+  case corruptLicense
+  /// container file does not contain the path to package file
+  case invalidContainerFile
+  /// license file does not match the book_vault_id we sent while requesting license
+  case invalidLicense
+  /// package file does not contain the required package endpoint
+  case invalidPackageFile
+  /// A dependency deallocated prematurely
+  case prematureDeallocation
+  /// user cancelled downlaod by pressing the cancel button
+  case userCancelledDownload
+  /// other error (either failed download or failure to write asset)
+  case other(Error)
+}
+
+/// `Task aggregator and synchronizer`
+///
+/// For a group of tasks such as [taskA, taskB, taskC, taskD], the execution order would be taskA(), taskB(),
+/// taskC(), taskD().
+///
+/// Tasks run synchronously so that taskD won't execute until taskC succeeds, and taskC won't execute until
+/// taskB succeeds.
+///
+/// At any point in time when a task fails, subsequent tasks don't get called. e.g. if taskB fails, taskC and taskD
+/// will never get called.
+///
+///       let aggregator = NYPLAxisTaskAggregator()
+///
+///       let taskA = NYPLAxisTask() { task in
+///         // some code that determines whether the task should succeed or fail
+///         task.succeeded() or task.failed(with error)
+///        }
+///
+///       let taskB = NYPLAxisTask() { task in
+///       // some code that determines whether the task should succeed or fail
+///         task.succeeded() or task.failed(with error)
+///        }
+///
+///       aggregator
+///         .addTasks([taskA, taskB])
+///         .run()
+///         .onCompletion { result in
+///           switch result {
+///             case .success:
+///             // Do something with success
+///             case .failure(let error):
+///             // Do something with success
+///           }
+///
+class NYPLAxisTaskAggregator {
+  private var tasks: [NYPLAxisTask] = []
+  private var completion: ((Result<Bool, NYPLAxisError>) -> Void)?
+  private var alreadyCompleted: Result<Bool, NYPLAxisError>?
+  private var cancelled: Bool = false
+  
+  
+  /// Add a task to the list of tasks
+  func addTask(_ task: NYPLAxisTask) -> NYPLAxisTaskAggregator {
+    self.tasks.append(task)
+    return self
+  }
+  
+  /// Add tasks to the list of tasks
+  func addTasks(_ tasks: [NYPLAxisTask]) -> NYPLAxisTaskAggregator {
+    self.tasks.append(contentsOf: tasks)
+    return self
+  }
+  
+  /// Start executing tasks
+  @discardableResult
+  func run() -> NYPLAxisTaskAggregator {
+    guard !cancelled else {
+      return self
+    }
+    
+    guard !tasks.isEmpty else {
+      alreadyCompleted = .success(true)
+      completion?(.success(true))
+      return self
+    }
+    
+    tasks.removeFirst().execute { (result) in
+      switch result {
+      case .success:
+        self.taskSucceeded()
+      case.failure(let error):
+        self.taskFailed(with: error)
```

Comment:
The class is retained until book download is finished/cancelled/failed. Once that's done, the instance is deallocated.

Reply:
did you notice anything unusual / cycles warnings / etc  in the memory graph debugger? It's not clear from here who's breaking the cycle so it might be worth checking.

Reply:
I did not notice any warnings